jax_test3.py

```diff
@@ -1,57 +1,117 @@
+    """
+    sequence = _derv_sequence(orders)
+    functions = [fun]
+    for i, order in enumerate(sequence):
+        functions.append(grad(functions[i], (order)))
+    return functions[-1](*variables)
 
-                for i in range(6):
-                    g_E_calc = grad(E_calc, i)
-                    for a in range(6):
-                        if i <= a:
-                            gg_E_calc = grad(g_E_calc, a)
-                            eval_E.append(gg_E_calc(coordinates[atom_a,0],coordinates[atom_a,1],coordinates[atom_a,2],coordinates[atom_b,0],coordinates[atom_b,1],coordinates[atom_b,2],charges[atom_a],charges[atom_b]))
-                            # print(f"{i},{a} {eval_E}")
-    return jnp.array(eval_E)
 
+def E_NN(N, Zs, *Rs):
+    if len(Rs) != N * 3:
+        raise RuntimeError(f"Must have {3*N} coordinates, got {len(Rs)}")
 
-coordinates = jnp.array([[1.0,2.0,3.0],[2.0,3.0,2.0],[1.0,2.0,4.0]])
-charges = [1.0,1.0,1.0]
-#E_list = E_calc(coordinates,charges)
-#print(E_list) 
-#var = 3
+    if len(Zs) != N:
+        raise RuntimeError(f"Must have {N} charges, got {len(Zs)}")
 
+    e_nn = 0
+    for A in range(N):
+        for B in range(A):
+            d_AB = jnp.sqrt(
+                (Rs[3 * A] - Rs[3 * B]) ** 2
+                + (Rs[3 * A + 1] - Rs[3 * B + 1]) ** 2
+                + (Rs[3 * A + 2] - Rs[3 * B + 2]) ** 2
+            )
+            e_nn += Zs[A] * Zs[B] / d_AB
 
-#gg_E_calc = grad(E_calc, 0)
-#testo = gg_E_calc(1.0,2.0,3.0,1.0,3.0,2.0,1.0,1.0)
-#print(testo)
+    return e_nn
 
-grad_E = get_hessian(jnp.array([[1.0,2.0,3.0],[2.0,1.0,3.0],[3.0,1.0,2.0]]),[1.0,1.0,1.0])
+
+def get_grad(coordinates, charges):
+    gradient = []
+    natoms = len(charges)
+    state = np.zeros(
+        2 + 3 * natoms, dtype=int
+    )  # makes an array that derv can work with
+    for item in range(2, len(state)):
+        state[item] = 1  # change the array to the wanted derivation for the hessian
+        eval_E = derv(
+            E_NN,
+            [charges.size, charges, *(coordinates.reshape(3 * natoms).tolist())],
+            state,
+        )
+        gradient.append(eval_E)
+        state[item] = 0  # reset array
+
+    gradient = np.array(gradient)
+    gradient = gradient.reshape((natoms, 3))
+    return gradient
+
+
+# this will generate the full hessian
+def get_hessian(coordinates, charges):
+    natoms = len(charges)
+    state = np.zeros(2 + 3 * natoms, dtype=int)
+    hessian = np.zeros(
+        (natoms, 3) * 2
+    )  # reshaping the hessian gave errors so here's created a priori
+    # these will keep track whether we are on the x y or z coordinate
+    i_xyz = 0
+    a_xyz = 0
+    for i in range(2, len(state)):
+        state[i] += 1  # change the item state
+        for a in range(2, len(state)):
+            state[a] += 1
+            eval_E = derv(
+                E_NN,
+                [charges.size, charges, *(coordinates.reshape(3 * natoms).tolist())],
```

Comment:
I think the `tolist` is not actually necessary and just using `*coordinates.reshape(3*natoms)` will work too. Note that the reshaping can be done once outside the loop and it should help with efficiency:
```python
flat_coords = coordinates.reshape(3*natoms)
for i in range(2, len(state)):
     ...
```